docs/topics/tasks.txt

```diff
@@ -0,0 +1,418 @@
+        )
+
+
+The return value of the decorator is a :class:`django.tasks.Task` instance.
+
+.. note::
+
+    All arguments and return values must be JSON serializable, and round-trip
+    as the same types.
+
+    This means complex types like model instances, as well as many built-in types
+    like ``datetime`` and ``tuple`` cannot be used in Tasks without additional
+    conversion.
+
+The ``task`` decorator accepts a few keyword arguments to customize the Task:
+
+* ``priority``: The priority of the Task. Higher numbers will be executed sooner.
+
+* ``queue_name``: The name of the queue the Task will be executed on
+
+* ``backend``: The name of the backend this Task must use (as defined in
+  :setting:`TASKS`).
+
+* ``enqueue_on_commit``: Whether the Task is enqueued when the current transaction
+  commits successfully, or enqueued immediately.
+
+These arguments correspond to attributes on the created
+:class:`Task <django.tasks.Task>`.
+
+Additionally, ``task`` accepts a ``takes_context`` argument, which controls whether
+the task function accepts context about its execution. See :ref:`task-context` for details.
+
+By convention, Tasks should be defined in a ``tasks.py`` file, however this is
+not enforced.
+
+.. _task-context:
+
+Task context
+------------
+
+Sometimes, the running task may need to know context about how it was enqueued,
+and how it is being executed. This can be accessed by taking a ``context`` argument,
+which is an instance of :class:`django.tasks.TaskContext`.
+
+To receive the task context as an argument to your task function, pass ``takes_context``
+when defining it::
+
+    from django.core.mail import send_mail
+    from django.tasks import task
+
+
+    @task(takes_context=True)
+    def email_users(context, emails, subject, message):
+        return send_mail(
+            subject=subject, message=message, from_email=None, recipient_list=emails
+        )
+
+The context has the following attributes:
+
+* :attr:`task_result <django.tasks.TaskContext.task_result>`
+* :attr:`attempt <django.tasks.TaskContext.attempt>`
+
+Modifying tasks
+---------------
+
+Before enqueueing Tasks, it may be necessary to modify certain parameters of the Task.
+For example, to give it a higher priority than it would normally.
+
+A ``Task`` instance cannot be modified directly. Instead, a modified instance
+can be created with the :meth:`using <django.tasks.Task.using>` method, leaving
+the original as-is::
+
+    email_users.priority  # 0
+
+    email_users.using(priority=10).priority  # 10
+
+``using`` allows modifying the following attributes:
+
+* :attr:`priority <django.tasks.Task.priority>`
+* :attr:`backend <django.tasks.Task.backend>`
+* :attr:`queue_name <django.tasks.Task.queue_name>`
+* :attr:`run_after <django.tasks.Task.run_after>`
+
+``run_after`` may also be provided as a :class:`timedelta <datetime.timedelta>`,
+which is used relative to the current time (when ``using`` is called), or a
```

Comment:
I think it would be nice to have a specific example setting `run_after` here to illustrate how it can be used. Help with this would be appreciated

Reply:
I think the help is to delete it - taking `timedelta` gets complex, as it's hard to understand what it's relative to (ie the time when `.using` is called, when `.enqueue` is called, and which timezone is used. I elected to leave it up to the user to handle instead, so this can be removed entirely.